core/transcoder.go

```diff
@@ -55,34 +56,92 @@ func NewLocalTranscoder(workDir string) Transcoder {
 	return &LocalTranscoder{workDir: workDir}
 }
 
+type nvSegResult struct {
+	*TranscodeData
+	error
+}
+
+type nvSegData struct {
+	session  *ffmpeg.Transcoder
+	fname    string
+	profiles []ffmpeg.VideoProfile
+	res      chan *nvSegResult
+}
+
 type NvidiaTranscoder struct {
-	workDir string
-	device  string
+	device  *segStack
 	session *ffmpeg.Transcoder
 }
 
-func (nv *NvidiaTranscoder) Transcode(job string, fname string, profiles []ffmpeg.VideoProfile) (*TranscodeData, error) {
-	// Set up in / out config
-	in := &ffmpeg.TranscodeOptionsIn{
-		Fname:  fname,
-		Accel:  ffmpeg.Nvidia,
-		Device: nv.device,
+type segStack struct {
+	segs []*nvSegData
+	mu   *sync.Mutex
+	cv   *sync.Cond
+	gpu  string
+}
+
+func newSegStack(gpu string) *segStack {
+	mu := &sync.Mutex{}
+	return &segStack{
+		segs: []*nvSegData{},
+		mu:   mu,
+		cv:   sync.NewCond(mu),
+		gpu:  gpu,
 	}
-	opts := profilesToTranscodeOptions(nv.workDir, ffmpeg.Nvidia, profiles)
+}
 
-	// Do the Transcoding
-	res, err := nv.session.Transcode(in, opts)
-	if err != nil {
-		return nil, err
+func (ss *segStack) push(seg *nvSegData) {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+	ss.segs = append(ss.segs, seg)
+	ss.cv.Broadcast()
+	if monitor.Enabled {
+		monitor.GPUBacklog(ss.gpu, len(ss.segs))
```

Comment:
Could there be a race condition here?

1. `pop()` locks `ss.mu`
2. `pop()` unlocks `ss.mu` in `ss.cv.Wait()` and pauses
3. `push()` locks `ss.mu` and appends to `ss.segs` s.t. the length is N + 1
4. `push()` invokes `ss.cv.Broadcast()`
5. `pop()` resumes from `ss.cv.Wait()` and locks `ss.mu`
6. `pop()` removes from `ss.segs` s.t. the length is N
7. `push()` calls `monitor.GPUBacklog()` with N instead of N + 1

Reply:
Yeah, that was actually a bit intentional - wanted to signal the broadcaster ASAP. Not sure if the condvar will release the lock immediately, but I hope it does because there's quite a bit more that happens as the metrics are being recorded. I'd rather give the system an opportunity to execute a pop / transcode if it can (even though it may not always do so).

What you're describing is an example of the [ABA problem](https://en.wikipedia.org/wiki/ABA_problem), where intermediate values are not recorded. In this case I think it's OK; the length is still correct at the moment it's being inserted into the metrics system. At step 7, the backlog is indeed N due to the `pop` in steps 5/6.

Can you think of a case where this might be problematic? Only thing I can think of is if there was a really long delay in re-obtaining the lock after `ss.cv.Broadcast()` for some reason. In the worst case, sustained `push` calls may cause metric to fall behind by at most one segment per `push` - but these pushes can't really make progress either without also obtaining the lock. (And vice versa with recording stack reductions after a `pop`.)

Reply:
Got it - that seems reasonable to me.

Could we add a comment mentioning that its possible for the metric recording might miss intermediate values just for clarification for a reader?

Reply:
Maybe I'm missing something here, but I don't see `Broadcast` function releasing lock internally - that means that step 6 will not execute until `push` returned.
@j0sh and if we want for it to work as you've described, we should not lock inside `push` function.

Reply:
Ah yeah I think @darkdarkdragon is right.

> Not sure if the condvar will release the lock immediately

When `Wait()` wakes up, it will lock the mutex before returning. But, the mutex is still locked at this point from step 3. So, `Wait()` will be paused until `push()` unlocks the mutex.

Reply:
and, btw, we can do
```golang
func (ss *segStack) push(seg *nvSegData) {
	ss.segs = append(ss.segs, seg)
        l := len(ss.segs)
	ss.cv.Broadcast()
	if monitor.Enabled {
		monitor.GPUBacklog(ss.gpu, l)
	}
}
```
this should fix it?

Reply:
Yeah, I should have just looked at the [source code](https://golang.org/src/sync/cond.go). So we actually need to hold the lock to protect all accesses to `ss.segs`, including during `len` . We could do something like:

```
lock
append seg
unlock
broadcast
lock
update metric with new length
unlock
```

Anyone have a preference for the above approach? I'm generally in favor of "less code" - keeping the current approach (but if shuffling code around gives us a small optimization win, I'll take it.) Too bad that's not the case here, but open to implementing it.

Reply:
Why not
```golang
func (ss *segStack) push(seg *nvSegData) {
	ss.mu.Lock()
        ss.segs = append(ss.segs, seg)
        l := len(ss.segs)
        ss.mu.Unlock()
	ss.cv.Broadcast()
	if monitor.Enabled {
		monitor.GPUBacklog(ss.gpu, l)
	}
}
```
?

Reply:
In that case we would be inserting the wrong length if `pop` was called before updating in the GPU backlog. We need to hold the lock during the metrics update.

Reply:
I thought it would just the behaviour you described in `Yeah, that was actually a bit intentional ` comment, no?

Reply:
> Yeah, that was actually a bit intentional

So that was referring to putting the metrics update *after* the broadcast, because I thought that might mean the broadcast would release the lock and maybe execute the transcode before updating the metric, then re-acquire the lock prior to returning from the `broadcast` call.

Since that's not actually the case, it doesn't really matter where the metrics update goes, so it's really just a matter of readability in this case. Do you think it's more readable if the metrics are updated before or after the broadcast? I can see how putting the metrics first might prevent some later confusion over the semantics of `broadcast`. But I'm also lazy enough that I'll leave this as-is unless someone asks for it :)